### PR TITLE
Add IPageFactory and IPageActivator

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

**Comment by pranavkm** [comment]
Redoing this PR because I borked https://github.com/aspnet/Mvc/pull/5536.

cc @rynowak

File: src/Microsoft.AspNetCore.Mvc.RazorPages/IPageActivatorProvider.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages
+{
+    /// <summary>
+    /// Provides methods to create a Razor page.
+    /// </summary>
+    public interface IPageActivatorProvider
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/IPageActivatorProvider.cs]: 😻

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageActivator.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (page == null)
+            {
+                throw new ArgumentNullException(nameof(page));
+            }
+
+            (page as IDisposable)?.Dispose();
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageActivator.cs]: You can also just do this check in `CreateDisposer` and then it doesn't even need to do anything at runtime.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageFactory.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            var modelType = actionDescriptor.ModelTypeInfo?.AsType() ?? actionDescriptor.PageTypeInfo.AsType();
+            var propertyActivator = new RazorPagePropertyActivator(
+                actionDescriptor.PageTypeInfo.AsType(),
+                modelType,
+                _modelMetadataProvider,
+                _propertyAccessors);
+
+            return (context) =>
+            {
+                var page = activatorFactory(context) as Page;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageFactory.cs]: This should be a c-style cast

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageFactory.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        }
+
+        public Action<PageContext, object> CreatePageDisposer(CompiledPageActionDescriptor descriptor)
+        {
+            if (descriptor == null)
+            {
+                throw new ArgumentNullException(nameof(descriptor));
+            }
+
+            return _pageActivator.CreateDisposer(descriptor);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageFactory.cs]: Let's make sure it's possible/allowed for these to return null if they aren't needed.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageFactory.cs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            return (context) =>
+            {
+                var page = activatorFactory(context) as Page;
+                page.PageContext = context;
+                propertyActivator.Activate(page, context);
+                return page;
+            };
+        }
+
+        public Action<PageContext, object> CreatePageDisposer(CompiledPageActionDescriptor descriptor)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageFactory.cs]: virtual?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageActivator.cs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+                throw new ArgumentException(Resources.FormatPropertyOfTypeCannotBeNull(
+                    nameof(actionDescriptor.PageTypeInfo),
+                    nameof(actionDescriptor)),
+                    nameof(actionDescriptor));
+            }
+
+            return CreatePageFactory(pageTypeInfo);
+        }
+
+        public virtual Action<PageContext, object> CreateDisposer(CompiledPageActionDescriptor actionDescriptor)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageActivator.cs]: Releaser?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvokerCacheEntry.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        }
+
+        public CompiledPageActionDescriptor ActionDescriptor { get; }
+
+        public Func<PageContext, object> PageFactory { get; }
+
+        public Action<PageContext, object> ReleasePage { get; }
+
+        Func<PageContext, IFilterMetadata[]> FilterProvider { get; }
+    }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvokerCacheEntry.cs]: 😻

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvokerProvider.cs [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+                {
+                    current = new InnerCache(actionDescriptors.Version);
+                    _currentCache = current;
+                }
+
+                return current;
+            }
+        }
+
+        public void OnProvidersExecuted(ActionInvokerProviderContext context)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvokerProvider.cs]: This should be in `OnProvidersExecuting`

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageFilterFactoryProvider.cs [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Microsoft.AspNetCore.Mvc.Abstractions;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.AspNetCore.Mvc.Internal;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages.Internal
+{
+    public static class PageFilterFactoryProvider
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageFilterFactoryProvider.cs]: Is this a copy of the FilterCache code?

I'm ok for now if you want to move forward with this PR and de-dupe in a follow up. My expectation is that we'll need want to de-dupe some of this, nothing about filters is really specific to pages.

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageFilterFactoryProviderTest.cs [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Abstractions;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.AspNetCore.Mvc.Internal;
+using Microsoft.AspNetCore.Routing;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages.Internal
+{
+    public class PageFilterFactoryProviderTest
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageFilterFactoryProviderTest.cs]: This really seems like a lot of stuff here that isn't specific to pages

Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Internal/PageFilterFactoryProviderTest.cs]: Yeah, it's copied from ControllerActionInvokerCacheTest. I'll de-dup this as part of the follow up.

**Pull request closed without merging**